include/xsimd/config/xsimd_arch.hpp

```diff
@@ -1,114 +1,175 @@
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
 #ifndef XSIMD_ARCH_HPP
 #define XSIMD_ARCH_HPP
 
-namespace xsimd {
-
-// forward declaration
-template<class T, size_t N> class batch;
-
-namespace arch {
-
-struct sse {
-  template<class T>
-  using batch = xsimd::batch<T, 128 / ( 8 * sizeof(T))>;
-  static constexpr size_t alignment = 8;
-};
-
-struct sse2 : sse {
-  static constexpr size_t alignment = 16;
-};
-
-struct sse3 : sse2 {
-};
-
-struct sse4_1 : sse3 {
-};
-
-struct sse4_2 : sse4_1{
-};
-
-struct avx {
-  template<class T>
-  using batch = xsimd::batch<T, 256 / ( 8 * sizeof(T))>;
-  static constexpr size_t alignment = 32;
-};
-
-// FIXME: unsure of that one
-struct fma3 : avx {
-};
-
-struct avx2 {
-  template<class T>
-  using batch = xsimd::batch<T, 256 / ( 8 * sizeof(T))>;
-  static constexpr size_t alignment = 32;
-};
-
-struct avx512 {
-  template<class T>
-  using batch = xsimd::batch<T, 512 / ( 8 * sizeof(T))>;
-  static constexpr size_t alignment = 64;
-};
-
-struct neon64 {
-  template<class T>
-  using batch = xsimd::batch<T, 512 / ( 8 * sizeof(T))>;
-  static constexpr size_t alignment = 32;
-};
-
-struct neon {
-  template<class T>
-  using batch = xsimd::batch<T, 512 / ( 8 * sizeof(T))>;
-  static constexpr size_t alignment = 16;
-};
-
-struct scalar {
-  template<class T>
-  using batch = xsimd::batch<T, 1>;
-  static constexpr size_t alignment = sizeof(void*);
-};
-
-template <class T, class InstructionSet>
-using batch = typename InstructionSet::template batch<T>;
+#include "xsimd_instruction_set.hpp"
+
+namespace xsimd
+{
+
+    // forward declaration
+    template<class T, size_t N>
+    class batch;
+
+    namespace arch
+    {
+        struct sse
+        {
+            template<class T>
+            using batch = xsimd::batch<T, 128 / ( 8 * sizeof(T))>;
+            static constexpr size_t alignment = 8;
+        };
+
+        struct sse2 : sse
+        {
+            static constexpr size_t alignment = 16;
+        };
+
+        struct sse3 : sse2
+        {
+        };
+
+        // Intel specific
+        struct sse4_1 : sse3
+        {
+        };
+
+        // Intel specific
+        struct sse4_2 : sse4_1
+        {
+        };
+
+        // AMD specific
+        struct sse4a : sse3
+        {
+        };
+
+        struct avx
+        {
+            template<class T>
+            using batch = xsimd::batch<T, 256 / ( 8 * sizeof(T))>;
+            static constexpr size_t alignment = 32;
+        };
+
+        struct fma3 : avx
+        {
+        };
+
+        // AMD specific (very few old processors)
+        struct fma4 : avx
+        {
+        };
+
+        // AMD specific (very few old processors)
+        struct xop : fma4
+        {
+        };
+
+        struct avx2
```

Comment:
@serge-sans-paille should avx2 inherit from avx?

Reply:
I'd say.